File: benchmark/autolifter/single-pass/mts.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mts = \xs: List. (fix (\f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0}
  | cons {h, t} ->
    let res = f t in
      {max (+ res.2 h) res.1, + res.2 h}
  end
) xs).1;

main = single_pass mts;

File: ml-benchmark/autolifter/single-pass/mts.ml
type l = Nil of unit | Cons of int * l

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let mts xs =
  let rec f xs =
    match xs with
    | Nil _ -> (0, 0)
    | Cons (h, t) ->
        let res = f t in
        match res with
        | (a1, a2) -> (max (a2 + h) a1, a2 + h)
  in
  let res = f xs in
  match res with
  | (a1, _) -> a1

let program = single_pass mts